### PR TITLE
make CEL admission controller code consumable

### PR DESCRIPTION
**Pull request by maxsmythe**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR makes the CEL Admission Controller code consumable by non-API-server clients by:

* Making the types returned by validation queries public, allowing callers to access the results of a policy check
* Separates version conversion from validation, removing the need to mock API-server-specific features when they are unavailable

/sig api-machinery
/cc @cici37 @jpbetz 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by alexzielenski** [comment]
related #115258

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go [hunk 1]
```diff
@@ -293,7 +294,14 @@ func (c *celAdmissionController) Validate(
 				}
 			}
 
-			decisions, err := bindingInfo.validator.Validate(a, o, param, matchKind)
+			versionedAttr, err := whgeneric.NewVersionedAttributes(a, matchKind, o)
+			if err != nil {
+				wrappedErr := fmt.Errorf("failed to convert object version: %w", err)
+				addConfigError(wrappedErr, definition, binding)
+				continue
+			}
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:302]: I think we can move this outside the binding loop. Seems like we are converting the object for every binding of the definition.

Another future optimization we might like to make is to share converted objects among definitions which match against the same kind. Maybe through some local `map[GVK]VersionedAttributes` or something

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/interface.go [hunk 2]
```diff
@@ -21,13 +21,14 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/generic"
 )
 
 // Validator defines the func used to validate the cel expressions
 // matchKind provides the GroupVersionKind that the object should be
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/interface.go:28]: Should update comment

**Review by alexzielenski — CHANGES_REQUESTED** [review]
lgtm. minor comments

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/interface.go (hunk 2 above)
Comment (maxsmythe) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/interface.go:28]: Oops, thanks! updated.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go (hunk 1 above)
Comment (maxsmythe) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller.go:302]: Good idea with only converting once, where possible.

Moving the actual logic outside the binding loop may cause unnecessary conversions when no binding matches, so I tried to split the difference.

**Review by alexzielenski — APPROVED** [review]
/approve
/label tide/merge-method-squash

lgtm will give cici some time to look

**Comment by cici37** [comment]
Thanks for the PR. It looks quite similar as https://github.com/kubernetes/kubernetes/pull/115258. We generally tend to not to export stuff for external usage since we don't guarantee the consistency and ideally we would love the info returned is used across all clients to avoid confusion. Would you mind providing further info on why this is needed? Thank you ^^
cc @jpbetz for your thoughts.

**Pull request merged**